### PR TITLE
Editor defaults for themes with no editor styles and when user disables theme styles

### LINKED ISSUE #35730: Editor styles are applied inconsistently for themes that do not support theme.json
### Description

If a theme does not support theme.json, editor styles are not applied on some themes, and text is rendered in the browser's default font.

This appears to have been introduced in #35424, when the check for `WP_Theme_JSON_Resolver_Gutenberg::theme_has_support()` was removed.

In testing a few different themes, it does appear there are theme-specific variables at play as well, and not simply whether or not the theme supports theme.json. How and when the theme declares its editor styles may be a factor (ie  `add_editor_style()` vs `wp_enqueue_style(`) and/or `init` vs `enqueue_block_editor_assets` for example)

### Step-by-step reproduction instructions

1. Activate a theme such as [Nucleare](https://wordpress.org/themes/nucleare/), which happens to be impacted
2. With Gutenberg v11.7 active, create a new post. Note the browser default font
3. Close the editor, and roll back to Gutenberg v11.6.
4. Create another new post, and note that the theme's editor styles are applied.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by oandregal**
Fixes https://github.com/WordPress/gutenberg/issues/35730
Follow up to https://github.com/WordPress/gutenberg/pull/35424 and https://github.com/WordPress/gutenberg/pull/34439

In Gutenberg 11.5, we added some defaults for themes that didn't enqueue any editor styles, see https://github.com/WordPress/gutenberg/pull/34439

In Gutenberg 11.7, we made every theme to have some styles in the editor: we provided at least the core presets. See https://github.com/WordPress/gutenberg/pull/35424 As an unintended side effect of this change, there are no themes that don't provide editor styles anymore, hence we didn't provide any editor defaults for them.

This PR adds back some defaults for themes that don't provide any editor styles.

### Make sure it works with "theme styles" deactivated

- I've tested this with TwentyTwentyOne, TT1-blocks, EmptyTheme, and Nucleare. Can't test with TwentyTwenty because it enqueues some things directly, no with the editor styles API.
- Go to the editor and in the "preferences" section disable "themes styles".
- Verify that editor styles look the same across different themes.
- Verify that presets work still as expected.

**Nucleare**

| This branch | v11.6 (last worked fine)  |
|---|---|
|  no-theme-styles-nucleare |  nucleare-no-theme-styles-v11-6 |

**Empty theme**

| This branch | v11.6 (last worked fine)  | 
|---|---|
|  no-theme-styles-empty-theme |  no-theme-styles-empty-theme-v11-6-1 |

**TT1 Blocks**

| This branch | v11.6 (last worked fine)  | 
|---|---|
| no-theme-styles-tt1-blocks | no-theme-styles-tt1-blocks-v11-6 |
 
**TwentyTwentyOne**

| This branch | v11.6 (last worked fine)  | 
|---|---|
| no-theme-styles-twenty-twenty-one | no-theme-styles-twenty-twenty-one-v11-6 |

### Make sure it works with "theme styles" activated

- I've tested this with TwentyTwentyOne, TT1-blocks, EmptyTheme, and Nucleare. Can't test with TwentyTwenty because it enqueues some things directly, no with the editor styles API.
- Go to the editor and in the preferences section enable "theme styles".
- Verify that editor styles look proper for the same themes above.
- Verify that presets work still as expected.

**Nucleare**

| This branch | v11.6 (last worked fine)  |
|---|---|
| theme-styles-nucleare |  nucleare-theme-styles-v11-6 |

**Empty theme**

| This branch | v11.6 (last worked fine)  | 
|---|---|
|  theme-styles-empty-theme |  theme-styles-empty-theme-v11-6 | 

**TT1 Blocks**

| This branch | v11.6 (last worked fine)  | 
|---|---|
|  theme-styles-tt1-blocks | theme-styles-tt1-blocks-v11-6 |
 
**TwentyTwentyOne**

| This branch | v11.6 (last worked fine)  | 
|---|---|
| theme-styles-twenty-twenty-one |  theme-styles-twenty-twenty-one-v11-6 |

### CODE REVIEW

File: lib/global-styles.php [hunk 1]
```diff
@@ -131,10 +131,14 @@ function_exists( 'gutenberg_is_edit_site_page' ) &&
 	}
 
 	if ( 'other' === $context ) {
-		$block_styles  = array( 'css' => gutenberg_experimental_global_styles_get_stylesheet( $consolidated, 'block_styles' ) );
+		$block_styles  = array(
+			'css'    => gutenberg_experimental_global_styles_get_stylesheet( $consolidated, 'block_styles' ),
+			'toSkip' => true,
```
Comment (oandregal) [inline lib/global-styles.php]: I'm not sure I like this so I've created an alternative PR at https://github.com/WordPress/gutenberg/pull/35737

**Comment by jasmussen** [comment]
Thanks for the followup. What's a good way to test? And do you personally prefer this PR or #35737?

File: packages/edit-post/src/editor.js [hunk 2]
```diff
@@ -161,7 +161,12 @@ function Editor( {
 	] );
 
 	const styles = useMemo( () => {
-		return hasThemeStyles && settings.styles?.length
+		// settings.styles can have styles that core adds.
+		// We don't want these to count as "theme styles".
+		const filteredStyles = settings.styles?.length
+			? settings.styles.filter( ( style ) => ! style.toSkip )
```
Comment (youknowriad) [inline packages/edit-post/src/editor.js]: So we're excluding these styles from the styles passed to the block editor. Why are we adding them in the backend in the first place then?

Comment (youknowriad) [reply packages/edit-post/src/editor.js]: Oh I think I understand now: "preset styles" are included in `settings.styles` and we only want to check for the styles that comes from the `theme.json styles` or custom editor styles here. 

I wonder if we should maybe expand on the `__unstableType` instead and clarify what each stylesheet is about, the "theme", "core" existing types don't seem to be enough anymore.

we need more granular things:

 - "core": Any core style that is loaded for all themes
 - "theme presets"
 - "theme styles" (theme.json or add_editor_styles calls)

It also seems like "theme presets" should also be included in the fallback in addition to `settings.defaultEditorStyles` to ensure the presets do work even if the user chooses to not user "theme styles" in its editor.

Comment (oandregal) [reply packages/edit-post/src/editor.js]: I saw your comment on the [other thread](https://github.com/WordPress/gutenberg/pull/35737#discussion_r732674783) and now understand this better. Will continue to iterate on this PR.

**Comment by oandregal** [comment]
> Thanks for the followup. What's a good way to test? And do you personally prefer this PR or #35737?

Initially, I thought #35737 was a better approach but after [some clarifications of the use cases to cover](https://github.com/WordPress/gutenberg/pull/35737#discussion_r732674783), I've closed that one and will focus on this direction instead. It needs a few tweaks, I'll ping you again when it's ready.

File: lib/global-styles.php (hunk 1 above)
Comment (oandregal) [reply lib/global-styles.php]: After some [clarifications](https://github.com/WordPress/gutenberg/pull/35737#discussion_r732674783) about the use cases to cover, I'm closing 35737 in favor of this one.

File: packages/edit-post/src/editor.js (hunk 2 above)
Comment (oandregal) [reply packages/edit-post/src/editor.js]: So, essentially we want the ability to remove theme styles in the post editor, for which they need to be enqueued separately. We need this:

1. theme.json styles (core)
2. editor styles provided via add_editor_styles (theme)
3. theme.json styles (theme)
4. theme.json styles (user)
5. preset classes (core, theme, user)
6. preset css vars for the sidebar, are not processed by .editor-styles-wrapper (core, theme, user)

So when the user disables "theme styles" we recalculate styles without 2 & 3. Perhaps we also skip 1 as well? Does this look correct to you?

Comment (oandregal) [reply packages/edit-post/src/editor.js]: This use case is new for me, and it's making me reconsider how we do some things.: would the styles for editors be simpler if we let them access the raw data so they can calculate the stylesheets themselves?

In addition to what I've shared for the post editor above (the 6 stylesheet plus the `__experimentalSettings` as well), this is what we currently send to the site editor:

- `__experimentalGlobalStylesUserEntityId` for theme.json user data
- `__experimentalGlobalStylesBaseConfig` with theme.json data (core+theme)
- `__experimentalSettings` theme.json settings (core+theme+user)
- editor styles provided via add_editor_styles

What if we just sent these four pieces of data instead:

- theme.json core
- theme.json theme
- theme.json user
- editor styles provided via add_editor_styles

Comment (youknowriad) [reply packages/edit-post/src/editor.js]: Based on the list above, I don't see why we should split 1, 2, and 4? It's fine if these are combined. and on the list you shared, the list from 1 to 4 are the things we want to exclude from the post editor if the user chooses to. So basically we only need to split presets and styles.

Comment (oandregal) [reply packages/edit-post/src/editor.js]: Don't we want to keep user styles (4) in that case?

Comment (youknowriad) [reply packages/edit-post/src/editor.js]: No, I don't think so, for me 4 is just the equivalent of custom CSS editor styles and when toggling the "theme styles", the user expect the same design across all themes.

File: packages/edit-post/src/editor.js [hunk 3]
```diff
@@ -161,9 +161,22 @@ function Editor( {
...
+		const themeStyles = [];
+		const presetStyles = [];
+		settings.styles.forEach( ( style ) => {
+			if ( ! style.__unstableType || style.__unstableType === 'theme' ) {
+				themeStyles.push( style );
+			} else {
+				presetStyles.push( style );
+			}
+		} );
+		const defaultEditorStyles = [
```
Comment (oandregal) [inline packages/edit-post/src/editor.js:173]: This is not working for themes that don't provide styles yet (Nucleare). The reason is styles provided by the plugin at https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php#L999

Comment (oandregal) [reply packages/edit-post/src/editor.js:173]: Also, need to investigate why TwentyTwenty still loads the body background color when theme styles are deactivated (the other themes I've tested work fine).

File: lib/class-wp-theme-json-gutenberg.php [hunk 4]
```diff
@@ -995,19 +995,6 @@ private function get_block_classes( $style_nodes ) {
 				$selector_duotone = self::scope_selector( $metadata['selector'], $metadata['duotone'] );
 				$block_rules     .= self::to_ruleset( $selector_duotone, $declarations_duotone );
 			}
-
-			if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
```
Comment (oandregal) [inline lib/class-wp-theme-json-gutenberg.php]: This is going to make the existing tests for styles fail. I need to fix them.

File: lib/compat/wordpress-5.9/spacing-styles.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * @param array $settings The existing editor settings.
+ * @return array The filtered editor settings.
+ */
+function gutenberg_load_spacing_styles_in_editor( $settings ) {
+	if ( ! WP_Theme_JSON_Resolver_Gutenberg::theme_has_support() ) {
+		return $settings;
+	}
+
+	$styles  = 'body { margin: 0; }';
```
Comment (oandregal) [inline lib/compat/wordpress-5.9/spacing-styles.php]: These can be extracted to a dedicated CSS file, like the default styles.

**Comment by oandregal** [comment]
@youknowriad what are your thoughts on the current direction?

The main thing is that the "spacing styles" (align, block gap) can't be part of the block styles generated as they are today: we need the block styles to be empty for themes that don't provide any editor styles.

File: packages/edit-post/src/editor.js (hunk 3 above)
Comment (oandregal) [reply packages/edit-post/src/editor.js:173]: The TwentyTwenty theme appears to enqueue some styles itself and not use editor styles ([see](https://github.com/WordPress/wordpress-develop/blob/f9279f497ca894c3959289affc39dffa5f2352eb/src/wp-content/themes/twentytwenty/functions.php#L420)). It's not a good theme for testing this PR then.

File: lib/compat/wordpress-5.9/spacing-styles.php (hunk 5 above)
Comment (oandregal) [reply lib/compat/wordpress-5.9/spacing-styles.php]: To avoid making this PR too big I can do this in a follow-up if folks want to do it.

**Comment by oandregal** [comment]
@jasmussen @chad1008 @youknowriad I've pushed all the changes and tested this across a number of themes. Left also test instructions in the issue description. It seems to work fine in my testing.

**Comment by chad1008** [comment]
This tests well for me across different themes, with and without theme styles active. Thanks @oandregal!

File: lib/global-styles.php [hunk 6]
```diff
@@ -9,43 +9,15 @@
...
- * @param WP_Theme_JSON_Gutenberg $tree Input tree.
- * @param string                  $type Type of stylesheet. It accepts 'all', 'block_styles', 'css_variables', 'presets'.
+ * @param WP_Theme_JSON_Gutenberg $tree  Input tree.
+ * @param array                   $types Which styles to load. It accepts 'css_variables', 'block_classes', 'preset_classes'. By default, it'll load all.
  *
  * @return string Stylesheet.
  */
-function gutenberg_experimental_global_styles_get_stylesheet( $tree, $type = null ) {
-	// Check if we can use cached.
-	$can_use_cached = (
```
Comment (oandregal) [inline lib/global-styles.php:19]: The cache in this method is dead code, it's never used, so we can remove it.

**Comment by jasmussen** [comment]
This appears to generally be working for me. 

One thing I'm noticing, is that the theme colors are not used here, these are the default WordPress blues:
Screenshot 2021-10-22 at 09 52 56

Matías encountered the same issue in #35662 ("Disabling theme editor styles in preferences makes some elements in the page load with default colors and not the current color scheme."), so it may be a separate issue. But at the same time, I see this in trunk:
Screenshot 2021-10-22 at 09 55 39

that is — the default styles with fonts and such are incorrect, but the colors are correct.

File: lib/class-wp-theme-json-gutenberg.php (hunk 4 above)
Comment (youknowriad) [reply lib/class-wp-theme-json-gutenberg.php]: Why did you have to remove this? This is not just for editor styles, it's also important for the frontend.

Comment (oandregal) [reply lib/class-wp-theme-json-gutenberg.php]: These styles are part of "block styles", hence, in the client, we can't distinguish the empty state (a theme with no "block classes").

Comment (oandregal) [reply lib/class-wp-theme-json-gutenberg.php]: Note that these weren't removed, but moved to another place `lib/compat/wordpress-5.9/spacing-styles.php`

**Comment by oandregal** [comment]
@jasmussen I'm struggling to see what the issue is. Tested with the TT1-blocks them in a few branches (including v11.6, where the issue this PR fixes was not happening) and they all look the same to me. Tried in Firefox and Chrome with no differences.

With theme styles:

| This branch | Trunk | v11.6 |
|---|---|---|
| tt1-blocks-with-theme-styles-branch |  tt1-blocks-with-theme-styles-trunk |  tt1-blocks-with-theme-styles-v11-6 |

No theme styles:

| This branch | Trunk | v11.6  |
|---|---|---|
| tt1-blocks-no-theme-styles-branch | tt1-blocks-no-theme-styles-trunk | tt1-blocks-no-theme-styles-v11-6 |

**Comment by jasmussen** [comment]
Go to your user profile and choose an admin color scheme other than the default one:
Screenshot 2021-10-22 at 11 45 28

Now observe that the color of the primary button remains `#007cba` (oldschool WordPress blue) even though it should be a color assigned by the color scheme. This _might_ be an existing issue with trunk and "no editor styles", and not a result of this PR, but wanted to just clarify that bit.

File: lib/class-wp-theme-json-gutenberg.php (hunk 4 above)
Comment (youknowriad) [reply lib/class-wp-theme-json-gutenberg.php]: > These styles are part of "block styles", hence, in the client, we can't distinguish the empty state (a theme with no "block classes").

Is a theme with theme.json ever a theme without editor styles? IMO no, so I think it's fine to consider them theme editor styles.

Comment (youknowriad) [reply lib/class-wp-theme-json-gutenberg.php]: The behavior of using the default editor styles in themes is specially tailored toward old themes that don't have any Gutenberg support (which is not the case for themes using theme.json)

**Comment by oandregal** [comment]
> This might be an existing issue with trunk and "no editor styles", and not a result of this PR, but wanted to just clarify that bit.

Oh, thanks for the instructions. I've checked v11.5, v11.6, `trunk` and this branch and all behave the same (colors don't represent the user color scheme), so it points to a different issue.

File: lib/class-wp-theme-json-gutenberg.php (hunk 4 above)
Comment (oandregal) [reply lib/class-wp-theme-json-gutenberg.php]: I was able to make this work properly while retaining those styles working as they are ([commit](https://github.com/WordPress/gutenberg/pull/35736/commits/c83eb9de85a910969e9e8ea40722c33652323b25)).

After what you said, I realized that those spacing styles should not be enqueued for themes without `theme.json`, neither in the editor or the frontend. While that is true for the frontend, we do enqueue the spacing styles in the editor in `trunk`. So this is a bug that we're also fixing with this PR ([here](https://github.com/WordPress/gutenberg/pull/35736/files#diff-0b8e98dead5542e95159bd3c482158111d3f16637e7eef9af13373358fa42dedR135)).

File: lib/class-wp-theme-json-gutenberg.php [hunk 7]
```diff
@@ -1261,30 +1261,34 @@ private static function get_setting_nodes( $theme_json, $selectors = array() ) {
...
+	 * @param array $types    Types of styles to load. Will load all by default. It accepts:
+	 *                         'css_variables': only CSS variables (presets & custom).
+	 *                         'block_classes': only block classes.
+	 *                         'preset_classes': only classes for presets.
+	 * @param array $origins A list of origins to include. By default it includes 'core', 'theme', and 'user'.
 	 *
 	 * @return string Stylesheet.
 	 */
-	public function get_stylesheet( $type = 'all', $origins = self::VALID_ORIGINS ) {
+	public function get_stylesheet( $types = array( 'css_variables', 'block_classes', 'preset_classes' ), $origins = self::VALID_ORIGINS ) {
```
Comment (youknowriad) [inline lib/class-wp-theme-json-gutenberg.php]: I wonder if `css_variables` should be a type. I mean we don't really know where these variables are used and it seems we should always render them anyway regardless of what types are passed.

File: lib/global-styles.php [hunk 8]
```diff
@@ -148,10 +108,39 @@ function_exists( 'gutenberg_is_edit_site_page' ) &&
...
+		foreach ( $new_presets as $new_style ) {
+			$style_css = gutenberg_experimental_global_styles_get_stylesheet( $consolidated, array( $new_style['css'] ) );
+			if ( '' !== $style_css ) {
+				$new_style['css']    = $style_css;
+				$new_global_styles[] = $new_style;
+			}
+		}
+
+		$new_block_classes = array(
+			'css'            => 'block_classes',
```
Comment (youknowriad) [inline lib/global-styles.php]: I'm unclear what this "type" is about?

Comment (youknowriad) [reply lib/global-styles.php]: I guess this is the theme styles generated from theme.json, the name is a bit confusing. It's not just about block classes, isn't it.

File: lib/class-wp-theme-json-gutenberg.php (hunk 7 above)
Comment (youknowriad) [reply lib/class-wp-theme-json-gutenberg.php]: I guess you want to avoid duplication in the applied stylesheets.

**Review by youknowriad — COMMENTED** [review]
This is looking good to me, the naming "block_classes" is a bit weird though.

File: lib/global-styles.php (hunk 8 above)
Comment (oandregal) [reply lib/global-styles.php]: So, these are the "types":

- `block_classes` => whatever is in `styles` plus the "spacing styles" we add
- `css_variables` => the css vars we generate for presets plus `settings.custom` (these need to be in scope for the sidebar UI controls, which is the main driver. They can't get rewritten by the `editor-styles-wrapper` while the other types can)
- `preset_classes` => the classes for the presets

File: lib/class-wp-theme-json-gutenberg.php (hunk 7 above)
Comment (oandregal) [reply lib/class-wp-theme-json-gutenberg.php]: responded https://github.com/WordPress/gutenberg/pull/35736#discussion_r734450254

File: lib/global-styles.php (hunk 8 above)
Comment (youknowriad) [reply lib/global-styles.php]: Maybe just "styles", "variables" and "presets".

Comment (oandregal) [reply lib/global-styles.php]: I perhaps can expand a bit more: for the preset UI controls to work, we need the CSS Custom Properties in the scope of the sidebar, as they use these values. They are currently scoped to `body` so we can't let the editor convert that to `.editor-styles-wrapper`.

Comment (oandregal) [reply lib/global-styles.php]: Those names work for me.

(sorry, GitHub didn't show me your comment)

Comment (oandregal) [reply lib/global-styles.php]: Updated the names.

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Pull request merged**